homeassistant/components/vacuum/mqtt.py

```diff
@@ -0,0 +1,539 @@
+                self._status = "Cleaning"
+            else:
+                self._status = "Stopped"
+
+            if topic == self._fan_speed_topic:
+                fan_speed = self._fan_speed_template\
+                    .async_render_with_possible_json_value(
+                        payload,
+                        error_value=None)
+                if fan_speed is not None:
+                    self._fan_speed = fan_speed
+
+            self.async_schedule_update_ha_state()
+
+        topics_set = {
+            self._battery_level_topic, self._charging_topic,
+            self._cleaning_topic, self._docked_topic, self._fan_speed_topic}
+        for topic in topics_set:
+            yield from self.hass.components.mqtt.async_subscribe(
+                topic, message_received, self._qos)
+
+    @property
+    def name(self):
+        """Return the name of the vacuum."""
+        return self._name
+
+    @property
+    def icon(self):
+        """Return the icon for the vacuum."""
+        return DEFAULT_ICON
+
+    @property
+    def should_poll(self):
+        """No polling needed for an MQTT vacuum."""
+        return False
+
+    @property
+    def is_on(self):
+        """Return true if vacuum is on."""
+        return self._cleaning
+
+    @property
+    def status(self):
+        """Return a status string for the vacuum."""
+        if self.supported_features & SUPPORT_STATUS == 0:
+            return
+
+        return self._status
+
+    @property
+    def fan_speed(self):
+        """Return the status of the vacuum."""
+        if self.supported_features & SUPPORT_FAN_SPEED == 0:
+            return
+
+        return self._fan_speed
+
+    @property
+    def fan_speed_list(self):
+        """Return the status of the vacuum."""
+        if self.supported_features & SUPPORT_FAN_SPEED == 0:
+            return []
+        return self._fan_speed_list
+
+    @property
+    def battery_level(self):
+        """Return the status of the vacuum."""
+        if self.supported_features & SUPPORT_BATTERY == 0:
+            return
+
+        return max(0, min(100, self._battery_level))
+
+    @property
+    def battery_icon(self):
+        """Return the battery icon for the vacuum cleaner."""
+        if self.supported_features & SUPPORT_BATTERY == 0:
+            return
+
+        return icon_for_battery_level(
+            battery_level=self.battery_level, charging=self._charging)
+
+    @property
+    def supported_features(self):
+        """Flag supported features."""
+        return self._supported_features
+
+    @asyncio.coroutine
+    def async_turn_on(self, **kwargs):
+        """Turn the vacuum on."""
+        if self.supported_features & SUPPORT_TURN_ON == 0:
+            return
+
+        mqtt.async_publish(self.hass, self._command_topic,
+                           self._payload_turn_on, self._qos, self._retain)
+        self._status = 'Cleaning'
+        self.async_schedule_update_ha_state()
```

Comment:
Since all of the methods have a very similar structure, an improvement (for the future) would be to add a helper function, that takes the support, payload and status as arguments. Then use the helper function in all these methods.

Reply:
Good idea